### PR TITLE
Add Python SDK

### PR DESCRIPTION
**Pull request by sbienkow-ninja**
##### Checklist

- [x] Tested in playground or other setup
- [ ] Screenshot (Grafana) from playground added to PR for 15+ minute run
- [x] Documentation is changed or added
- [ ] Tests and/or benchmarks are included
- [ ] Breaking changes

- - -
This change is [Reviewable](https://reviewable.io/reviews/fluxninja/aperture/1574)

### Summary by OpenAI

**Release Notes**

This pull request adds a Python SDK for Aperture, including documentation and examples. It also includes updates to the CircleCI configuration file, pre-commit hooks, and scripts to generate gRPC code from proto files. The changes also include updates to Java and Go SDKs, as well as modifications to the playground and example app.

- New Feature: Added Python SDK for Aperture with OpenTelemetry tracing and an OTLP exporter.
- Documentation: Added documentation for manually setting feature control points using the Aperture Python SDK and a link to the Python SDK in the list of supported languages.
- Refactor: Updated Java and Go SDKs to handle null responses and add null checks.
- Chore: Updates to CircleCI configuration and scripts to generate gRPC code and install pre-commit hooks.

### CODE REVIEW

File: sdks/aperture-java/lib/core/src/main/java/com/fluxninja/aperture/sdk/ApertureSDK.java [hunk 1]
```diff
@@ -82,16 +82,13 @@ public Flow startFlow(String controlPoint, Map<String, String> explicitLabels) {
...
       res = this.flowControlClient
           .withDeadlineAfter(timeout.toNanos(), TimeUnit.NANOSECONDS)
           .check(req);
     } catch (StatusRuntimeException e) {
       // deadline exceeded or couldn't reach agent - request should not be blocked
-      res = CheckResponse.newBuilder()
-          .setDecisionType(CheckResponse.DecisionType.DECISION_TYPE_ACCEPTED)
-          .build();
     }
     span.setAttribute(WORKLOAD_START_TIMESTAMP_LABEL, Utils.getCurrentEpochNanos());
```
[the remark this answers is not in the document]
Comment (sbienkow-ninja) [reply sdks/aperture-java/lib/core/src/main/java/com/fluxninja/aperture/sdk/ApertureSDK.java:94]: This was done to make Java SDK behave the same way as Go/JS/Python sdk.

File: sdks/aperture-java/lib/core/src/main/java/com/fluxninja/aperture/sdk/ApertureSDK.java [hunk 2]
```diff
@@ -115,14 +112,13 @@ public TrafficFlow startTrafficFlow(String path, AttributeContext attributes) {
...
+    com.fluxninja.generated.envoy.service.auth.v3.CheckResponse res = null;
     try {
       res = this.envoyAuthzClient
           .withDeadlineAfter(timeout.toNanos(), TimeUnit.NANOSECONDS)
           .check(req);
     } catch (StatusRuntimeException e) {
       // deadline exceeded or couldn't reach agent - request should not be blocked
-      res = TrafficFlow.successfulResponse();
     }
     span.setAttribute(WORKLOAD_START_TIMESTAMP_LABEL, Utils.getCurrentEpochNanos());
```
[the remark this answers is not in the document]
Comment (sbienkow-ninja) [reply sdks/aperture-java/lib/core/src/main/java/com/fluxninja/aperture/sdk/ApertureSDK.java:124]: This was done to make Java SDK behave the same way as Go/JS/Python sdk.

[the remark this answers is not in the document]
Comment (sbienkow-ninja) [reply sdks/aperture-java/lib/core/src/main/java/com/fluxninja/aperture/sdk/ApertureSDK.java:124]: This was done to make Java SDK behave the same way as Go/JS/Python sdk.

File: cmd/sdk-validator/main.go [hunk 3]
```diff
@@ -38,7 +38,11 @@ var (
...
 
 func init() {
-	logger = log.NewLogger(log.GetPrettyConsoleWriter(), log.DebugLevel.String())
+	logLevel, logLevelSet := os.LookupEnv("LOG_LEVEL")
+	if !logLevelSet {
+		logLevel = log.DebugLevel.String()
+	}
+	logger = log.NewLogger(log.GetPrettyConsoleWriter(), logLevel)
 	log.SetGlobalLogger(logger)
 }
```
[the remark this answers is not in the document]
Comment (sbienkow-ninja) [reply cmd/sdk-validator/main.go:48]: This is an internal tool and if level is incorrect, we want to fail.

**Review by IridiumOxide — APPROVED** [review]
Java part looks good, but we should work out what to do with indentation changes (merging this will cause the changed files to have different indentation levels than files not included in this PR)

**Comment by sbienkow-ninja** [comment]
> Java part looks good, but we should work out what to do with indentation changes (merging this will cause the changed files to have different indentation levels than files not included in this PR)

My vote is to add something like [spotless](https://github.com/diffplug/spotless) to gradle to have single source of truth for Java code, regardless of user IDE. For python I added pre-commit to format the code :)

**Review by kklimonda-fn — APPROVED** [review]
> Reviewed 32 of 43 files at r1, 11 of 11 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/fluxninja/aperture/1574)* status: all files reviewed, 15 unresolved discussions (waiting on @github-actions[bot] and @sbienkow-ninja)

___
*[`docs/content/get-started/integrations/flow-control/sdk/python/manual.md` line 20 at r2](https://reviewable.io/reviews/fluxninja/aperture/1574#-NQjhpyq9od1W8p839j1:-NQjhpyq9od1W8p839j2:bg3tj8k) ([raw file](https://github.com/fluxninja/aperture/blob/725e405497140baa2b8425b593940f99431b7433/docs/content/get-started/integrations/flow-control/sdk/python/manual.md#L20)):*
> ```Markdown
> -->
> 
> To do so, first create an instance of ApertureClient:
> ```

Nit: It's nice to have some introduction, and not just start page with "To do so" (so what? :D)

File: sdks/aperture-java/lib/core/src/main/java/com/fluxninja/aperture/sdk/ApertureSDK.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (harjotgill) [reply sdks/aperture-java/lib/core/src/main/java/com/fluxninja/aperture/sdk/ApertureSDK.java:71]: @openai This was done to make Java SDK behave the same way as Go/JS/Python sdk.

File: sdks/aperture-java/lib/core/src/main/java/com/fluxninja/aperture/sdk/ApertureSDK.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (harjotgill) [reply sdks/aperture-java/lib/core/src/main/java/com/fluxninja/aperture/sdk/ApertureSDK.java:100]: @openai This was done to make Java SDK behave the same way as Go/JS/Python sdk.

File: cmd/sdk-validator/main.go (hunk 3 above)
Comment (harjotgill) [reply cmd/sdk-validator/main.go:48]: @openai This is an internal tool and if level is incorrect, we want to fail.

File: sdks/aperture-java/lib/core/src/main/java/com/fluxninja/aperture/sdk/ApertureSDK.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (harjotgill) [reply sdks/aperture-java/lib/core/src/main/java/com/fluxninja/aperture/sdk/ApertureSDK.java:100]: @openai This was done to make Java SDK behave the same way as Go/JS/Python sdk.

File: docs/content/get-started/integrations/flow-control/sdk/python/manual.md
Comment (harjotgill) [inline docs/content/get-started/integrations/flow-control/sdk/python/manual.md:1]: @openai can you please improve this documentation change?

Comment (harjotgill) [reply docs/content/get-started/integrations/flow-control/sdk/python/manual.md:1]: Please rewrite it by fixing grammar etc.

File: docs/content/get-started/integrations/flow-control/sdk/python/manual.md [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+For more context on how to use Aperture Python SDK to set feature Control
+Points, you can take a look at the [example app][example] available in our
+repository.
+
+<!--- TODO: Change to pypi package once it is published. --->
+
+[pythonsdk]: https://github.com/fluxninja/aperture/tree/main/sdks/aperture-py
+[example]:
+  https://github.com/fluxninja/aperture/tree/main/sdks/aperture-py/example
```
Comment (harjotgill) [inline docs/content/get-started/integrations/flow-control/sdk/python/manual.md:88]: @openai can you please rewrite or improve this documentation change?

File: docs/content/get-started/integrations/flow-control/sdk/python/manual.md
Comment (harjotgill) [inline docs/content/get-started/integrations/flow-control/sdk/python/manual.md:1]: @openai can you please improve this documentation diff to make it more concise?

File: docs/content/get-started/integrations/flow-control/sdk/python/manual.md (hunk 4 above)
Comment (harjotgill) [inline docs/content/get-started/integrations/flow-control/sdk/python/manual.md:88]: @openai can you please improve this documentation diff to make it more concise?

Comment (harjotgill) [reply docs/content/get-started/integrations/flow-control/sdk/python/manual.md:88]: rewrite this line -
`To do so, first create an instance of ApertureClient:`

File: docs/content/get-started/integrations/flow-control/sdk/python/manual.md
Comment (harjotgill) [inline docs/content/get-started/integrations/flow-control/sdk/python/manual.md:1]: @openai can you please summarize this file?

File: sdks/aperture-java/lib/core/src/main/java/com/fluxninja/aperture/sdk/ApertureSDK.java
Comment (harjotgill) [inline sdks/aperture-java/lib/core/src/main/java/com/fluxninja/aperture/sdk/ApertureSDK.java:1]: @openai please help add documentation-style comments on top of all methods in this file and provide the updated file.

Comment (harjotgill) [reply sdks/aperture-java/lib/core/src/main/java/com/fluxninja/aperture/sdk/ApertureSDK.java:1]: Please add Java-style comments.

Comment (harjotgill) [reply sdks/aperture-java/lib/core/src/main/java/com/fluxninja/aperture/sdk/ApertureSDK.java:1]: Please provide updated file in your response beloe.

File: sdks/aperture-py/aperture_sdk/client.py
Comment (harjotgill) [inline sdks/aperture-py/aperture_sdk/client.py:1]: @openai Please add documentation as comments on each class and function of this file.

Comment (harjotgill) [reply sdks/aperture-py/aperture_sdk/client.py:1]: Please generate the new code with the documentation.

Comment (harjotgill) [inline sdks/aperture-py/aperture_sdk/client.py:1]: @openai What are classes and their methods in this file? Provide a bullet-point list.

Comment (harjotgill) [reply sdks/aperture-py/aperture_sdk/client.py:1]: Please provide the developer-style documentation for each function in a table. E.g. document the parameters and returned values.

File: sdks/aperture-py/aperture_sdk/flow.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    def __enter__(self: TFlow) -> TFlow:
+        return self
+
+    def __exit__(self, exc_type, _exc_value, _traceback) -> None:
+        if self._ended:
+            return
+        if exc_type is not None:
+            self.end(FlowStatus.Error)
+        else:
+            self.end(FlowStatus.OK)
```
Comment (harjotgill) [inline sdks/aperture-py/aperture_sdk/flow.py:77]: @openai can you please do a thorough code review of this file?

File: sdks/aperture-py/aperture_sdk/utils.py [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+
+
+async def run_fn(fn: Callable[..., TWrappedReturn], *args, **kwargs) -> TWrappedReturn:
+    """Run a function or coroutine."""
+    # We want to support both sync and async functions
+    # https://stackoverflow.com/questions/44169998/how-to-create-a-python-decorator-that-can-wrap-either-coroutine-or-function
+    if inspect.iscoroutinefunction(fn):
+        return await fn(*args, **kwargs)
+    else:
+        return fn(*args, **kwargs)
```
Comment (harjotgill) [inline sdks/aperture-py/aperture_sdk/utils.py:14]: @openai can you please do a thorough code review of this file?

File: scripts/generate_proto.sh [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+sed -i "s/^from aperture\..* import \([^ ]*_pb2\) as \([^ ]*\)$/from . import \1 as \2/" "${generated_py_files[@]}"
+generated_pyi_files_str="$(find "$python_gen" -type f -name '*.pyi')"
+readarray -t generated_pyi_files <<< "$generated_pyi_files_str"
+sed -i "s/^import grpc$/import grpc\nimport grpc.aio\nimport typing/" "${generated_pyi_files[@]}"
+sed -i "s/: grpc\.Server/: typing.Union[grpc.Server, grpc.aio.Server]/" "${generated_pyi_files[@]}"
+
+echo "Generating __init__.py files"
+find "$python_gen" -type d -exec touch "{}/__init__.py" \;
+
+echo "Done"
```
Comment (harjotgill) [inline scripts/generate_proto.sh:56]: @openai can you please do a thorough review of this file? Are there better ways to write this code?

File: sdks/aperture-py/aperture_sdk/flow.py
Comment (harjotgill) [inline sdks/aperture-py/aperture_sdk/flow.py:1]: @openai can you write unit tests for this file?

Comment (harjotgill) [reply sdks/aperture-py/aperture_sdk/flow.py:1]: Check for correctness based on your best judgement. Eg

Comment (harjotgill) [reply sdks/aperture-py/aperture_sdk/flow.py:1]: Please generate python code file for the unit tests

Comment (harjotgill) [inline sdks/aperture-py/aperture_sdk/flow.py:1]: @openai please generate a Python code that unit tests Flow class and it’s functions for correctness. You can use pytest framework.

**Pull request merged**